File: packages/local/internal/indexhost/native/syntax/record/interests.go
package record

import (
	"github.com/use-crux/crux/packages/local/internal/indexhost/native/syntax"
	"github.com/use-crux/crux/packages/local/internal/projectindex"
)

func CallInterests(input []projectindex.StaticCallInterest) []syntax.CallInterest {
	if len(input) == 0 {
		return nil
	}
	interests := make([]syntax.CallInterest, 0, len(input))
	for _, interest := range input {
		interests = append(interests, syntax.CallInterest{
			Name:       interest.Name,
			ImportFrom: append([]string(nil), interest.ImportFrom...),
			ConfigArg:  interest.ConfigArg,
			Properties: append([]string(nil), interest.Properties...),
			Callbacks:  callbackInterests(interest.Callbacks),
			Source:     interest.Source,
		})
	}
	return interests
}

func ConstructorInterests(input []projectindex.StaticConstructorInterest) []syntax.ConstructorInterest {
	if len(input) == 0 {
		return nil
	}
	interests := make([]syntax.ConstructorInterest, 0, len(input))
	for _, interest := range input {
		interests = append(interests, syntax.ConstructorInterest{
			Name:       interest.Name,
			ImportFrom: append([]string(nil), interest.ImportFrom...),
			ConfigArg:  interest.ConfigArg,
			Properties: append([]string(nil), interest.Properties...),
			Callbacks:  callbackInterests(interest.Callbacks),
			Source:     interest.Source,
		})
	}
	return interests
}

func callbackInterests(input []projectindex.StaticCallbackInterest) []syntax.CallbackInterest {
	if len(input) == 0 {
		return nil
	}
	interests := make([]syntax.CallbackInterest, 0, len(input))
	for _, interest := range input {
		interests = append(interests, syntax.CallbackInterest{
			Property: interest.Property,
			MaxDepth: interest.MaxDepth,
		})
	}
	return interests
}

File: packages/indexer/src/indexer/static-index/protocol/interests.ts
/**
 * Parser-interest fields shared by Static Index prepare and analyze plans.
 *
 * These fields intentionally mirror the syntax-record worker filters. The
 * Static Index compiler consumes compiler-owned manifest interests instead of
 * hardcoding first-party or extension call names.
 *
 * @module
 */

import { z } from 'zod'

/** Declared callback property retained for parser evidence slices. */
export const StaticIndexParserCallbackInterestSchema = z
  .object({
    property: z.string().min(1),
    maxDepth: z.number().int().nonnegative().optional(),
  })
  .strict()

/** Import-aware call filter carried from the static syntax plan. */
export const StaticIndexParserCallInterestSchema = z
  .object({
    name: z.string().min(1),
    importFrom: z.array(z.string().min(1)).optional(),
    configArg: z.number().int().nonnegative().optional(),
    properties: z.array(z.string().min(1)).optional(),
    callbacks: z.array(StaticIndexParserCallbackInterestSchema).optional(),
    source: z.string().min(1).optional(),
  })
  .strict()

/** Import-aware constructor filter carried from the static syntax plan. */
export const StaticIndexParserConstructorInterestSchema = StaticIndexParserCallInterestSchema

/** Parser filters that must stay identical across prepare and analyze. */
export const staticIndexParserInterestFields = {
  callNames: z.array(z.string().min(1)).optional(),
  callInterests: z.array(StaticIndexParserCallInterestSchema).optional(),
  constructorNames: z.array(z.string().min(1)).optional(),
  constructorInterests: z.array(StaticIndexParserConstructorInterestSchema).optional(),
  pruneNativeFactCallNames: z.array(z.string().min(1)).optional(),
} satisfies z.ZodRawShape

/** Parser callback interest. */
export type StaticIndexParserCallbackInterest = z.infer<typeof StaticIndexParserCallbackInterestSchema>

/** Parser call interest. */
export type StaticIndexParserCallInterest = z.infer<typeof StaticIndexParserCallInterestSchema>

/** Parser constructor interest. */
export type StaticIndexParserConstructorInterest = z.infer<typeof StaticIndexParserConstructorInterestSchema>
